### PR TITLE
doc: Generating lcov.info file

### PR DESCRIPTION
**Pull request by iamriajul**

### CODE REVIEW

**Comment by iamriajul** [comment]
@googlebot I signed it!

File: README.md [hunk 1]
```diff
@@ -61,3 +61,9 @@ collected. If `--sdk-root` is set, Dart SDK coverage will also be output.
 - `// coverage:ignore-line` to ignore one line.
 - `// coverage:ignore-start` and `// coverage:ignore-end` to ignore range of lines inclusive.
 - `// coverage:ignore-file` to ignore the whole file.
+
+#### Generating lcov.info file (used by codecov bash uploader)
+- Run test with coverage: `pub run test --coverage coverage`
+- Formatting to lcov.info: `pub global run coverage:format_coverage -i coverage/test -o coverage/lcov.info --lcov`
```
[the remark this answers is not in the document]
Comment (grouma) [reply README.md:67]: > @iamriajul -- thanks for the patch!
> 
> These are covered in the 'Collecting coverage' and 'Formatting coverage' sections above, though we should probably document the `--pretty-print` and `--lcov` options explicitly in the text of the format section.
> 
> @grouma if pub now includes `--coverage` built in, I wonder if we shouldn't just inject something like this patch right under the 'Tools' section, as a sort of 'preferred' way of using the package (i.e. via `pub run test --coverage`). Then the existing sections could be left for the few people who might want to do collection directly against the VM (e.g. tools authors).

pub doesn't include `--coverage` built in but `package:test` does. I'm not sure I follow your suggestion.

> 
> Alternatively, I do also like the idea of adding a section on integrating with various code coverage or CI services. We could just start with this one, and anyone else who'd like to add theirs could send a PR.
> 
> Something like:
> 
> ```gfm
> ### Integration with coverage services
> You can use this tooling to compute and upload code coverage to any service
> that accepts LCOV format files as input.
> 
> #### Codecov.io 
> ...
> 
> #### Coveralls
> ...
> ```
> 
> One last question for @grouma: Is the output of the `pub run test --coverage` the JSON blob? I almost wonder if the ideal answer isn't that we just create `pub coverage --lcov` and `pub coverage --pretty-print`. `pub run test --coverage` could then be deprecated, or do whatever it does now.

Yes `package:test` will output the JSON blob to the specified location provided with `--coverage`. We did this to be consistent with the `vm` coverage behavior which was implemented well before the web support. Note there should be one JSON blob file per platform. Ideally we want the coverage behavior in `package:test`  to be consistent with coverage support in `flutter test` but that requires a breaking change. This issue comes up every so often but hasn't been a high enough priority to deal with the churn of a breaking change in `package:test`.

cc @natebosch

[the remark this answers is not in the document]
Comment (cbracken) [reply README.md:67]: > pub doesn't include --coverage built in but package:test does. I'm not sure I follow your suggestion.

Sorry my wording was bad. If pub run test (which, as you say, is implemented in package:test) now supports a `--coverage` option built in, then we should consider just documenting that that is the preferred way to collect coverage, rather than the details currently in this readme. To be clear, the details currently in this readme should remain, but be primarily aimed developers of tooling that does some form of coverage collection, rather than at end users collecting coverage for their own projects.

> Yes package:test will output the JSON blob to the specified location provided with --coverage. We did this to be consistent with the vm coverage behavior which was implemented well before the web support. Note there should be one JSON blob file per platform. 

Ideally there'd just be no JSON blob. The history here is that the JSON blob _used_ to be used for backward compatibility with other (VM) tooling that hasn't existed since probably 2014 other than `format_coverage` and some internal tooling which may or may not exist any more. Ideally we'd just emit lcov and if there's any desire to keep the pretty-print functionality, built it off the LCOV instead. It's entirely possible that external tools have developed that rely on it at this point, so it probably needs to be supported for compatibility, but long-term it'd be nice to have test runs with `--coverage` emit.

> I almost wonder if the ideal answer isn't that we just create `pub run coverage --lcov` and `pub run coverage --pretty-print`. `pub run test --coverage` could then be deprecated, or do whatever it does now.

To be clear the suggestion here is sort of the inverse of just having `pub run test --coverage` emit LCOV. Users would run `pub run coverage --lcov`, which would do the setup, then delegate the test run vial package:test APIs, then emit in whichever format. 

I have no opinion on which is better. I imagine just having package:test emit LCOV when `--coverage` is set is least-effort other than dealing with backward compatibility concerns.

[the remark this answers is not in the document]
Comment (grouma) [reply README.md:67]: > Sorry my wording was bad. If pub run test (which, as you say, just invokes code in package:test) now supports a --coverage option built in, then we should consider just documenting that that is the preferred way to collect coverage, rather than the details currently in this readme. To be clear, the details currently in this readme should remain, but be primarily aimed developers of tooling that does some form of coverage collection, rather than at end users collecting coverage for their own projects. Does that make sense?

That makes sense to me.

> To be clear the suggestion here is sort of the inverse of just having pub run test --coverage emit LCOV. Users would run pub run coverage --lcov, which would do the setup, then delegate the test run via package:test APIs, then emit in whichever format the user selected.

This makes me nervous as `package:test` is no longer orchestrating things but technically `package:coverage`. It also may be weird from a dependency perspective and what a user puts in their `pubspec.yaml`.

> TL;DR: for the purposes of this patch, I'm suggesting we just document that we encourage users to collect coverage via pub run test --coverage, rather than via collect_coverage and that this change, therefore, be moved up near the top of this readme.

Makes sense to me.

[the remark this answers is not in the document]
Comment (greymouser) [reply README.md:67]: I just spent much too long searching the differences between flutter and dart test coverage invocation, in order to deploy a truly, fully dart package, and thankfully I stumbled upon this proposed patch.

@cbracken README.md does *not* give a clear example for collecting coverage from a test run. And the formatting line was also interesting. I don't know if README.md had this information in the past, but it's woefully inadequate at the moment. Your second comment seems to be aware of this -- so I just wanted to chime in that this issue hit me as well. Fixing it is too easy not to be high priority.

This is a fantastic, simple PR that would have saved me 30 minutes or so. If I could immediately merge it, I would.

[the remark this answers is not in the document]
Comment (Stargator) [reply README.md:67]: So far from my testing it does not seem `dart test` includes function-coverage or branch-coverage which are supported by coverage:test_with_coverage

**Comment by mosuem** [comment]
Closing this as it is stale. Feel free to reopen in case you still want to land it!
